rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,266 @@
+```
+
+The existing constant tensor cache feature is disabled by default which means
+zero capacity on both CPU and GPU. In some memory limited scenarios, users
+may only want to cache a part of constant tensors to improve performance, but
+don't want the cached capacity to grow infinitely and consume too much memory.
+In this document, we propose to introduce a constant tensor cache capacity
+control API to address above requests.
+
+## Related works
+
+### TensorRT
+
+Users can control the max workspace size used inside TensorRT through the
+`nvinfer1::IBuilderConfig::setMemoryPoolLimit` and
+`nvinfer1::IBuilderConfig::getMemoryPoolLimit` APIs. If the workspace size is
+not big enough, TensorRT will show warning message to suggest users to increase
+the workspace size. If OOM occurs, users could also reduce workspace memory
+consumption through this API. And users are suggested to set the workspace size
+as big as possible to achieve best performance. The default workspace size is
+set to device global memory size from v8.4 release.
+
+## Proposal
+
+### Dedicated cache for different engine kind
+In oneDNN Graph programming model, different graph can have different engine
+kind and simultaneously run on different engine, like cpu or gpu. Different kind
+of engine usually has different memory limitation. 
+
+So, we propose to use dedicated cache for different engine kind, and let users
+set different capacity for different engine kind separately.
+
+For multiple devices scenarios, like multi-GPU, whether to use dedicated cache
+for each device is still an open question, we will discuss it in later sections.
+
+### Unified cache for all backends
+We propose to implement an unified constant tensor cache in the oneDNN Graph
+interface layer. Each backend should use this interface to cache their all
+constant tensors into the unified cache. And the oneDNN Graph API users could
+control the capacity of the unified cache, no need to be aware of any backend
+specific things. This proposal is based on the following considerations:
+
+- In memory limited scenarios, we usually want to control the total memory
+  footprint, instead of the memory footprint of specific backend.
+- By the design of oneDNN Graph API, the API users don't need to be aware about
+  which backend is using. We don't like to break this design currently.
+- Backend selection is dynamic in mixed backend scenarios. Users can't
+  explicitly know which backend will run and how many partition will a backend
+  claims until graph partition stage done. So, it's hard to decide how many
+  memory a backend should use.
+
+### Capacity setting/getting frontend API
+Based on above considerations, we proposed to add the following two C++ API and
+corresponding C API to control the capacity of constant tensor cache:
+
+```cpp
+void set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, int64_t capacity);
+int64_t get_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind);
```

Comment:
Using `engine_kind` is not generic since the scenario when two GPUs with different allocators provided will not be covered. Using `engine` instead should provide all necessary information for the implementation.

`int64_t` type doesn't make any sense for capacity which mustn't be negative. There's only one choice - `size_t`.

Expecting users to call two APIs to enable constant cache is not a friendly behavior. It looks like that just setting capacity should be enough. If the value is greater than zero, it implies the cache is enabled. It means that former API seems to be obsolete and should be deprecated.

Reply:
Thanks for review~
>Using engine_kind is not generic since the scenario when two GPUs with different allocators provided will not be covered. Using engine instead should provide all necessary information for the implementation.

OK, I will consider it.
> int64_t type doesn't make any sense for capacity which mustn't be negative. There's only one choice - size_t.

Sorry for being careless, the implementation API has already changed to size_t, thanks for pointing it out. I will update the RFC.

>Expecting users to call two APIs to enable constant cache is not a friendly behavior. It looks like that just setting capacity should be enough. If the value is greater than zero, it implies the cache is enabled. It means that former API seems to be obsolete and should be deprecated.

We here just have only one API to enable constant cache which is `set_constant_tensor_cache_capacity`, and `get_constant_tensor_cache_capacity` is used for users to acquire the specific capacity.  The old API `set_constant_tensor_cache` will be deprecated in the future release.